### PR TITLE
Initiate the inner usage of `cfg_match` (Compiler)

### PR DESCRIPTION
**Pull request by c410-f3r**
cc #115585

Dogfood to test the implementation and remove dependencies.

### CODE REVIEW

File: compiler/rustc_data_structures/src/lib.rs [hunk 1]
```diff
@@ -6,45 +6,44 @@
...
 #![feature(unwrap_infallible)]
-#![feature(strict_provenance)]
-#![feature(ptr_alignment_type)]
-#![feature(macro_metavar_expr)]
-#![allow(rustc::default_hash_types)]
-#![allow(rustc::potential_query_instability)]
-#![deny(rustc::untranslatable_diagnostic)]
-#![deny(rustc::diagnostic_outside_of_impl)]
-#![allow(internal_features)]
-#![deny(unsafe_op_in_unsafe_fn)]
```
Comment (c410-f3r) [inline compiler/rustc_data_structures/src/lib.rs:41]: Alphabetical order

File: compiler/rustc_data_structures/src/marker.rs [hunk 2]
```diff
@@ -48,13 +49,9 @@ cfg_if!(
             [std::io::StdoutLock<'_>]
             [std::io::StderrLock<'_>]
         );
-        cfg_if!(
-            // Consistent with `std`
-            // `os_imp::Env` is `!Send` in these platforms
-            if #[cfg(any(unix, target_os = "hermit", target_os = "wasi", target_os = "solid_asp3"))] {
-                impl !DynSend for std::env::VarsOs {}
-            }
-        );
```
Comment (c410-f3r) [inline compiler/rustc_data_structures/src/marker.rs:57]: Unnecessary

File: compiler/rustc_span/src/lib.rs [hunk 3]
```diff
@@ -13,19 +13,20 @@
...
-#![feature(read_buf)]
+#![feature(min_specialization)]
+#![feature(negative_impls)]
 #![feature(new_uninit)]
-#![deny(rustc::untranslatable_diagnostic)]
-#![deny(rustc::diagnostic_outside_of_impl)]
-#![allow(internal_features)]
+#![feature(read_buf)]
+#![feature(round_char_boundary)]
+#![feature(rustc_attrs)]
```
Comment (c410-f3r) [inline compiler/rustc_span/src/lib.rs:29]: Alphabetical order

File: compiler/rustc_data_structures/src/lib.rs (hunk 1 above)
Comment (fbstj) [reply compiler/rustc_data_structures/src/lib.rs:41]: is it worth a [tidy-alphabetical](https://doc.rust-lang.org/stable/nightly-rustc/tidy/alphabetical/index.html) comment?

Comment (c410-f3r) [reply compiler/rustc_data_structures/src/lib.rs:41]: Done

**Comment by c410-f3r** [comment]
r? compiler

**Comment by c410-f3r** [comment]
Should be good now

**Pull request merged**